pkg/engine/common/base.go

```diff
@@ -26,15 +26,19 @@ import (
 )
 
 type Shared struct {
-	Headers    map[string]string
-	KnownFiles *files.KnownFiles
-	Options    *types.CrawlerOptions
+	Headers      map[string]string
+	KnownFiles   *files.KnownFiles
+	Options      *types.CrawlerOptions
+	InFlightUrls mapsutil.SyncLockMap[string, bool]
```

Comment:
If you don't need to hold a value but only keys, I'd recommend:
```
mapsutil.SyncLockMap[string, struct{}]
```
structs hold no pointers and increase performances